ufl/corealg/dag_traverser.py

```diff
@@ -0,0 +1,120 @@
+"""Base class for dag traversers."""
+
+from abc import ABC, abstractmethod
+from functools import singledispatchmethod, wraps
+
+from ufl.classes import Expr
+
+
+class DAGTraverser(ABC):
+    """Base class for dag traversers."""
+
+    def __init__(self, compress=True, vcache=None, rcache=None) -> None:
+        """Initialise."""
+        self._compress = compress
+        # Temporary data structures
+        # expr -> r = function(expr,...),  cache of intermediate results
+        self._vcache = {} if vcache is None else vcache
+        # r -> r,  cache of result objects for memory reuse
+        self._rcache = {} if rcache is None else rcache
+
+    def __call__(self, node: Expr, *args) -> Expr:
+        """Perform memoised DAG traversal with ``process`` singledispatch method.
+
+        Args:
+            node: `Expr` to start DAG traversal from.
+            args: arguments to the ``process`` singledispatchmethod.
+
+        Returns:
+            Processed `Expr`.
+
+        """
+        cache_key = (node, *args)
+        try:
+            return self._vcache[cache_key]
+        except KeyError:
+            result = self.process(node, *args)
+            # Optionally check if r is in rcache, a memory optimization
+            # to be able to keep representation of result compact
+            if self._compress:
+                result_ = self._rcache.get(result)
+                if result_ is None:
+                    # Cache miss: store in rcache
+                    self._rcache[result] = result
+                else:
+                    # Cache hit: Use previously computed object result_,
+                    # allowing result to be garbage collected as soon as possible
+                    result = result_
```

Comment:
Might this be cleaner?
```suggestion
                try:
                    result = self._rcache[result]
                except KeyError:
                    self._rcache[result] = result
```

Also this seems kind of like a `set` to me, not a `dict`.